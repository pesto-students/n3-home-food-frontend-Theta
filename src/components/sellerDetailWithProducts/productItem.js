import { Button, Col, Row, Typography, notification, Tag } from "antd";

import { React, useState, useEffect } from "react";
import { rupeeSign } from "utils/constant";
import { sessionId } from "utils/helpers";
import Image from "../image/image";
import "./sellerDetailWithProducts.css";
import { useDispatch } from "react-redux";
import { setSellerIdInCart } from "store/actions/index";
import { useTranslation } from "react-i18next";
import { updateCartItem } from "../utils/api";

const ProductItems = ({ products, savedCartItem, reloadCart, sellerId }) => {
  const Dispatch = useDispatch();
  const { t } = useTranslation();

  const { Title } = Typography;
  const [allProducts, setAllProducts] = useState([...products]);

  useEffect(() => {}, [savedCartItem]);

  useEffect(() => {
    setAllProducts([...products]);
  }, [products]);

  const addItems = (dish, key, method) => {
    if (!sessionId()) {
      notification.error({
        message: t("Message.Notification"),
        description: t("Message.PleaseLogin"),
        placement: "topRight",
      });
      return;
    }

    allProducts[key].isLoading = true;
    setAllProducts([...allProducts]);
    let currentProduct = savedCartItem.filter(
      (item) => item.productId.id === dish.productId
    );
    let cartItem = {};
    if (currentProduct.length > 0) {
      if (method === "add") {
        currentProduct[0].quantity = currentProduct[0].quantity + 1;
      }
      if (method === "sub") {
        currentProduct[0].quantity = currentProduct[0].quantity - 1;
      }

      cartItem = {
        productId: dish.productId,
        quantity: currentProduct[0].quantity,
        userId: sessionId(),
        price: currentProduct[0].price,
      };
    } else {
      cartItem = {
        productId: dish.productId,
        quantity: 1,
        userId: sessionId(),
        price: dish.price,
      };
    }

    updateCart(cartItem, key);
  };

  const updateCart = async (cartItem, key) => {
    const response = await updateCartItem(cartItem);
    if (response.status === 200) {
      allProducts[key].isLoading = false;
      setAllProducts([...allProducts]);
      Dispatch(setSellerIdInCart(sellerId));
      reloadCart();
    }
  };

  const getQuantity = (currentItem) => {
    try {
      let current = savedCartItem.filter(
        (item) => item.productId.id === currentItem.productId
      );
      if (current.length > 0) {
        return current[0].quantity;
      }
      return 0;
    } catch (e) {
      return 0;
    }
  };

  return (
    <Row>
      {allProducts.map((dish, key) => {
        return (
          <Col sm={24} xs={24} md={24} key={key} className="product-row">
            <Row>
              <Col md={6} sm={8} xs={8}>
                <Image url={dish.image} height="95" width="95"></Image>
              </Col>
              <Col md={18} sm={14} xs={14}>
                <Title level={4} className="dish-head">
                  {dish.name}{" "}
                  <span>
                    ({t("ProductItem.Only")} {dish.quantity}{" "}
                    {t("ProductItem.Left")})
                  </span>{" "}
                </Title>

                <Title level={5} className="dish-description">
                  {dish.description}
                </Title>

                <Row className="category-tags">
                  {dish.productCategory.map((category) => {
                    return <Tag color="processing">{category.name}</Tag>;
                  })}
                </Row>
                <Row justify="space-between" align="middle">
                  <p level={5}>
                    {rupeeSign} {dish.price}
                  </p>

                  {getQuantity(dish) === 0 ? (
                    <Button
                      loading={dish.isLoading}
                      onClick={() => addItems(dish, key, "add")}
                    >
                      {t("ProductItem.Add")}
                    </Button>
                  ) : (
                    <div>
                      <Button
                        loading={dish.isLoading}
                        onClick={() => addItems(dish, key, "sub")}
                      >
                        -
                      </Button>
                      <Button>{getQuantity(dish)}</Button>
                      <Button
                        loading={dish.isLoading}
                        onClick={() => addItems(dish, key, "add")}
                      >
                        +
                      </Button>
                    </div>
                  )}
                </Row>
              </Col>
            </Row>
          </Col>
        );
      })}
    </Row>
  );
};

export default ProductItems;
