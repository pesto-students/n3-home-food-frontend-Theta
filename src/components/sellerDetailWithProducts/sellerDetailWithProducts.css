.seller-banner{
    height: 300px;
    width: 100%;
    display: flex;
    background:#000000d9;
    color: #fff;
    padding: 40px;
}
.seller-banner h1,.seller-banner h5{
    color: #fff;
}
.image-and-detail{
    width: 100%;
}
.margin-10{
    margin-top: 10px;
}
.category-tabs{
    margin: 40px;
}
.product-row{
    margin:15px 0px;
    border-bottom: 1px solid #7777771c;
    padding-bottom:10px
}

.product-text{
    border-bottom: 1px solid #7777771c;
    padding-bottom: 15px;
    font-size: 20px !important;

}

.dish-head{
    font-size: 15px !important;
    margin: 0px !important;
    color: #777 !important;
}
.dish-description{
    font-size: 13px !important;
    margin: 0px !important;

    color: #777 !important;
}
.category-tags{
    margin:15px 0px;
}

.seller-details{
    margin-left: 20px;
}

@media screen and (max-width: 500px) {
    .seller-banner{
        height: auto;
        width: 100%;
        display: flex;
        background:#000000d9;
        color: #fff;
        padding: 0px
    }
    .seller-details{
           margin: 0px;
           align-items: center;
           text-align: center;
           padding: 10px 10px;
           justify-content: center;
    }
        }