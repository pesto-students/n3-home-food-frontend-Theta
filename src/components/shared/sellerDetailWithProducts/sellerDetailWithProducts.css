.seller-banner{
    height: 300px;
    width: 100%;
    display: flex;
    background:#000000d9;
    color: #fff;
    padding: 25px;
}
.seller-banner h1,.seller-banner h5{
    color: #fff;
}
.row{
    width: 100%;
}
.margin-10{
    margin-top: 10px;
}
.category-tabs{
    margin: 40px;
}
.product-row{
    margin:15px 0px;
    padding-bottom: 25px;
    border-bottom: 1px solid rgb(218, 196, 196);
}

.product-text{border-bottom: 1px solid #777;
    padding-bottom: 25px;

}