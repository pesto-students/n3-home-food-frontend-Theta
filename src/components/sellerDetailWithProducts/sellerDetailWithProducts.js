import { React, useEffect, useState, useCallback } from "react";
import { Row, Col, Typography, Rate, Card, Tabs } from "antd";
import "./sellerDetailWithProducts.css";
import CustomerNavbar from "../customerNavbar/customerNavbar";
import Image from "../image/image";
import ProductItems from "./productItem";
import SpinnerLoader from "../spinnerLoader/spinnerLoader";
import Cart from "../cart/cart";
import { getUser, sessionId, setPincode } from "utils/helpers";
import { withRouter } from "react-router-dom";
import Navbar from "../navbar/navbar";
import { useTranslation } from "react-i18next";
import DataNotFound from "components/dataNotFound/dataNotFound";
import { fetchSellerProfile, fetchCart } from "../utils/api";

const categoryList = ["All", "Breakfast", "Lunch", "Snack", "Dinner"];

const SellerDetailWithProducts = ({ match }) => {
  const { TabPane } = Tabs;
  const { t } = useTranslation();
  const sellerId = match.params.id;
  const { Title } = Typography;
  const [isLoading, setIsLoading] = useState(true);
  const [isCartLoad, setIsCartLoad] = useState(false);
  const [allProduct, setAllProduct] = useState(false);
  const [profile, setProfile] = useState({});
  const [tabSelected, setTabSelected] = useState(0);
  const [pageSize, setPageSize] = useState(5);
  const [alreadyInCart, setAlreadyInCart] = useState({
    items: [],
  });

  const getCurrentTab = (tab) => {
    setTabSelected(tab);
    if (tab === "0") {
      let products = [];
      profile.myProducts.forEach((element, index) => {
        if (index > pageSize) return;
        products.push(element);
      });
      setAllProduct([...products]);
      return;
    }
    let items = profile.myProducts.filter(
      (item) =>
        item.productCategory.filter((n) => n.name === categoryList[Number(tab)])
          .length > 0
    );
    let products = [];
    items.forEach((element, index) => {
      if (index > pageSize) return;
      products.push(element);
    });
    setAllProduct([...products]);
  };

  const getSellerProfile = useCallback(async () => {
    const response = await fetchSellerProfile(sellerId);
    if (response.status === 200) {
      setProfile(response.data[0]);
      let products = [];
      response.data[0].myProducts.forEach((element, index) => {
        if (index > pageSize) return;
        products.push(element);
      });
      setAllProduct([...products]);
      setIsLoading(false);
    }
  }, [sellerId, pageSize]);

  const getCart = async () => {
    if (sessionId()) {
      const response = await fetchCart(sessionId()).catch((e) => {
        setIsCartLoad(true);
        let cart = {
          items: [],
          subTotal: 0,
        };
        setAlreadyInCart(cart);
      });
      setAlreadyInCart(response.data);
      setIsCartLoad(true);
    } else {
      setIsCartLoad(true);
    }
  };

  const fetchMoreProducts = useCallback(async () => {
    let pageLimit = pageSize + 5;
    setPageSize(pageLimit);
    if (tabSelected === "0") {
      let products = [];

      profile.myProducts.forEach((element, index) => {
        if (index > pageLimit) return;
        products.push(element);
      });
      setAllProduct([...products]);
      return;
    }
  }, [pageSize, profile, tabSelected]);

  useEffect(() => {
    getSellerProfile();
    getCart();
  }, [getSellerProfile]);

  const uddatePincode = (code) => {
    setPincode(code);
  };

  return (
    <>
      {getUser() ? (
        <CustomerNavbar updatePincode={uddatePincode}></CustomerNavbar>
      ) : (
        <Navbar callBack={uddatePincode} />
      )}
      {isLoading === true ? (
        <Row justify="center">
          <SpinnerLoader />
        </Row>
      ) : (
        <Row>
          <Col md={24}>
            <div className="seller-banner">
              <Row className="image-and-detail">
                <Col md={6} sm={24} xs={24}>
                  <Image url={profile.image} height="200" width="100%"></Image>
                </Col>
                <Col md={16} sm={24} xs={24} className="seller-details">
                  <Title level={1}>{profile.name}</Title>
                  <Title level={5}>{profile.description}</Title>
                  <Rate defaultValue={profile.rating}></Rate>
                </Col>
              </Row>
            </div>

            <Row justify="center" className="margin-10">
              <div className="category-tabs">
                <Tabs onChange={getCurrentTab}>
                  <TabPane tab={t("Landing.All")} key="0" />
                  <TabPane tab={t("Landing.breakfast")} key="1" />
                  <TabPane tab={t("Landing.lunch")} key="2" />
                  <TabPane tab={t("Landing.snacks")} key="3" />
                  <TabPane tab={t("Landing.dinner")} key="4" />
                </Tabs>
              </div>
            </Row>

            <Row>
              <Col md={6} sm={2} xs={2}></Col>
              <Col md={12} sm={22} xs={22}>
                <Title level={3} className="product-text">
                  {t("Cart.Products")}
                </Title>

                {allProduct.length === 0 && (
                  <DataNotFound text={t("Header.notFound")}></DataNotFound>
                )}

                {isCartLoad && (
                  <ProductItems
                    products={allProduct}
                    reloadCart={getCart}
                    savedCartItem={alreadyInCart.items}
                    sellerId={sellerId}
                    fetchMoreProducts={fetchMoreProducts}
                  />
                )}
              </Col>
              <Col md={6} xs={24} sm={24}>
                <Card>
                  <Title className="product-text" level={3}>
                    {t("Cart.Cart")}
                  </Title>
                  {isCartLoad && (
                    <Cart reloadCart={getCart} alreadyInCart={alreadyInCart} />
                  )}
                </Card>
              </Col>
            </Row>
          </Col>
        </Row>
      )}
    </>
  );
};

export default withRouter(SellerDetailWithProducts);
