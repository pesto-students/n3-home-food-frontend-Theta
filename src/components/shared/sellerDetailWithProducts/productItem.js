import { Button, Col, Row, Typography, notification, Tag } from "antd";
import "antd/dist/antd.css";
import { React, useState } from "react";
import axios from "../../../utils/axios";
import { baseUrl, rupeeSign } from "../../../utils/constant";
import { sessionId } from "../../../utils/helpers";
import Image from "../image/image";
import "./sellerDetailWithProducts.css";
import { useDispatch } from "react-redux";
import { setSellerIdInCart } from "../../../store/actions/index";

const ProductItems = ({ products, savedCartItem, reloadCart, sellerId }) => {
  const Dispatch = useDispatch();

  const { Title } = Typography;
  const [isLoading, setIsLoding] = useState(false);

  const addItems = (dish, method) => {
    if (!sessionId()) {
      notification.error({
        message: "Notification",
        description: "Please Login to add into cart",
        placement: "topRight",
      });
      return;
    }

    setIsLoding(true);
    let currentProduct = savedCartItem.filter(
      (item) => item.productId === dish.productId
    );

    let cartItem = {};
    if (currentProduct.length > 0) {
      if (method === "add") {
        currentProduct[0].quantity = currentProduct[0].quantity + 1;
      }
      if (method === "sub") {
        currentProduct[0].quantity = currentProduct[0].quantity - 1;
      }
      console.log("cartItem", currentProduct[0]);

      cartItem = {
        productId: dish.productId,
        quantity: currentProduct[0].quantity,
        userId: sessionId(),
        price: currentProduct[0].price,
      };
    } else {
      cartItem = {
        productId: dish.productId,
        quantity: 1,
        userId: sessionId(),
        price: dish.price,
      };
    }

    updateCart(cartItem);
  };

  const updateCart = (cartItem) => {
    axios
      .post(`${baseUrl}/cart`, cartItem)
      .then((result) => {
        setIsLoding(false);
        Dispatch(setSellerIdInCart(sellerId));

        reloadCart();
      })
      .catch((err) => console.error(err));
  };

  const getQuantity = (currentItem) => {
    try {
      console.log("savedCartItem", savedCartItem);
      let current = savedCartItem.filter(
        (item) => item.productId === currentItem.productId
      );
      if (current.length > 0) {
        return current[0].quantity;
      }
      console.log("current", currentItem);
      return 0;
    } catch (e) {
      return 0;
    }
  };

  return (
    <>
      <Row>
        {products.map((dish, key) => {
          return (
            <Col sm={24} xs={24} md={24} key={key} className="product-row">
              <Row>
                <Col md={2}>
                  <Image url={dish.image} height="40" width="40"></Image>
                </Col>
                <Col md={22}>
                  <Title level={4}>
                    {dish.name} <span>(Only {dish.quantity} Left)</span>
                  </Title>
                  {dish.productCategory.map((category) => {
                    return <Tag color="processing">{category.name}</Tag>;
                  })}
                  <Title level={5}>{dish.description}</Title>
                  <Row justify="space-between" align="middle">
                    <p level={5}>
                      {rupeeSign} {dish.price}
                    </p>

                    {getQuantity(dish) === 0 ? (
                      <Button
                        loading={isLoading}
                        onClick={() => addItems(dish, "add")}
                      >
                        Add
                      </Button>
                    ) : (
                      <div>
                        <Button
                          loading={isLoading}
                          onClick={() => addItems(dish, "sub")}
                        >
                          -
                        </Button>
                        <Button>{getQuantity(dish)}</Button>
                        <Button
                          loading={isLoading}
                          onClick={() => addItems(dish, "add")}
                        >
                          +
                        </Button>
                      </div>
                    )}
                  </Row>
                </Col>
              </Row>
            </Col>
          );
        })}
      </Row>
    </>
  );
};

// const mapStateToProps = (state) => {
//   return {
//     cartItems: state.myCart,
//   };
// };

// export default connect(mapStateToProps)(ProductItems);
export default ProductItems;
