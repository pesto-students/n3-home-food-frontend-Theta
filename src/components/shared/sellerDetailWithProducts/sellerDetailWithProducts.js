import "antd/dist/antd.css";
import { React, useEffect, useState, useCallback } from "react";
import { Row, Col, Typography, Rate, Card } from "antd";
import "./sellerDetailWithProducts.css";
import CustomerNavbar from "../customerNavbar/customerNavbar";
import Image from "../image/image";
import CustomTabs from "../Tabs/Tabs";
import ProductItems from "./productItem";
import { baseUrl, rupeeSign } from "../../../utils/constant";
import axios from "../../../utils/axios";
import SpinnerLoader from "../spinnerLoader/spinnerLoader";
import Cart from "../cart/cart";
import { getUser, sessionId } from "../../../utils/helpers";
import { withRouter } from "react-router-dom";
import Navbar from "../navbar/navbar";
import { useTranslation } from "react-i18next";


const SellerDetailWithProducts = (props) => {
  const { t } = useTranslation();
  const sellerId = props.match.params.id;
  const { Title } = Typography;
  const [isLoading, setIsLoading] = useState(true);
  const [isCartLoad, setIsCartLoad] = useState(false);

  const [profile, setProfile] = useState({});
  const [alreadyInCart, setAlreadyInCart] = useState({
    items: [],
  });

  const getCurrentTab = (tab) => {
    console.log(tab);
    let items = profile.myProducts.filter((item) =>
      item.productCategory.filter((n) => n.name === tab)
    );
    console.log(items);
  };

  const getSellerProfile = useCallback(() => {
    axios
      .get(`${baseUrl}/sellers/get/getproducts?sellerid=${sellerId}`)
      .then((result) => {
        setProfile(result.data[0]);
        setIsLoading(false);
      })
      .catch((err) => console.error(err));
  }, [sellerId]);

  const getCart = () => {
    axios
      .get(`${baseUrl}/cart/${sessionId()}`)
      .then((result) => {
        setAlreadyInCart(result.data);
        setIsCartLoad(true);
      })
      .catch((err) => {
        setIsCartLoad(true);
        let cart = {
          items: [],
          subTotal: 0,
        };
        setAlreadyInCart(cart);

        console.error(err);
      });
  };

  useEffect(() => {
    getSellerProfile();
    getCart();
  }, [getSellerProfile]);

  return (
    <>
      {getUser() ? <CustomerNavbar></CustomerNavbar> : <Navbar />}
      {isLoading === true ? (
        <Row justify="center">
          <SpinnerLoader />
        </Row>
      ) : (
        <Row>
          <Col md={24}>
            <div className="seller-banner">
              <Row gutter={["30"]} className="row">
                <Col md={6}>
                  <Image url="" height="250" width="100%"></Image>
                </Col>
                <Col md={18}>
                  <Title level={1}>{profile.name}</Title>
                  <Title level={5}>{profile.description}</Title>
                  <Rate defaultValue={profile.rating}></Rate>
                  <Title className="margin-10" level={5}>
                    {t('Cart.Max Amount')} : {rupeeSign} {profile.max_amount}
                  </Title>
                </Col>
              </Row>
            </div>

            <Row justify="center" className="margin-10">
              <div className="category-tabs">
                <CustomTabs
                  currentTab={getCurrentTab}
                  list={["All", "Breakfast", "Lunch", "Snack", "Dinner"]}
                />
              </div>
            </Row>

            <Row>
              <Col md={6}></Col>
              <Col md={12}>
                <Title level={3} className="product-text">
                   {t('Cart.Products')}
                </Title>

                {isCartLoad && (
                  <ProductItems
                    products={profile.myProducts}
                    reloadCart={getCart}
                    savedCartItem={alreadyInCart.items}
                    sellerId={sellerId}
                  />
                )}
              </Col>
              <Col md={6}>
                <Card>
                  <Title level={3}> {t('Cart.Cart')} </Title>
                  <hr />
                  {isCartLoad && (
                    <Cart reloadCart={getCart} alreadyInCart={alreadyInCart} />
                  )}
                </Card>
              </Col>
            </Row>
          </Col>
        </Row>
      )}
    </>
  );
};

export default withRouter(SellerDetailWithProducts);
