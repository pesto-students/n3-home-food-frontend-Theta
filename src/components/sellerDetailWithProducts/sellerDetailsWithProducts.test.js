import * as React from "react";
import Enzyme, { shallow } from "enzyme";
import EnzymeAdapter from "enzyme-adapter-react-16";
import sellerDetailWithProducts from "./sellerDetailWithProducts";

Enzyme.configure({ adapter: new EnzymeAdapter() });

describe("Seller Card component", () => {
  let wrapper;
  test("render without error", () => {
    wrapper = shallow(<sellerDetailWithProducts/>);
    expect(wrapper.length).toBe(1);
  });

  test('should render correctly in "debug" mode', () => {
    const component = shallow(<sellerDetailWithProducts />);

    expect(component).toMatchSnapshot();
  });
});