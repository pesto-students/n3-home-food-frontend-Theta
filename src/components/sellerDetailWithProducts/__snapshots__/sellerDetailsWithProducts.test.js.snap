// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seller Card component should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;
